packages/@sanity/client/sanityClient.d.ts

```diff
@@ -0,0 +1,1784 @@
+     */
+    create(
+      name: string,
+      options?: {aclMode?: DatasetAclMode}
+    ): Promise<{datasetName: string; aclMode: DatasetAclMode}>
+
+    /**
+     * Edit a dataset with the given name
+     *
+     * @param name Name of the dataset to edit
+     * @param options New options for the dataset
+     */
+    edit(
+      name: string,
+      options: {aclMode?: DatasetAclMode}
+    ): Promise<{datasetName: string; aclMode: DatasetAclMode}>
+
+    /**
+     * Delete a dataset with the given name
+     *
+     * @param name Name of the dataset to delete
+     */
+    delete(name: string): Promise<{deleted: true}>
+
+    /**
+     * Fetch a list of datasets for the configured project
+     */
+    list(): Promise<{name: string; aclMode: DatasetAclMode}[]>
+  }
+
+  projects: {
+    /**
+     * Fetch a list of projects the authenticated user has access to
+     */
+    list(): Promise<SanityProject[]>
+
+    /**
+     * Fetch a project by project ID
+     *
+     * @param projectId ID of the project to fetch
+     */
+    getById(projectId: string): Promise<SanityProject>
+  }
+
+  users: {
+    /**
+     * Fetch a user by user ID
+     *
+     * @param id User ID of the user to fetch. If `me` is provided, a minimal response including the users role is returned.
+     */
+    getById<T extends 'me' | string>(
+      id: T
+    ): T extends 'me' ? Promise<CurrentSanityUser> : Promise<SanityUser>
+  }
+
+  auth: {
+    /**
+     * Fetch available login providers
+     */
+    getLoginProviders(): Promise<{providers: AuthProvider[]}>
+
+    /**
+     * Revoke the configured session/token
+     */
+    logout(): Promise<any>
+  }
+
+  /**
+   * Set up a listener that will be notified when mutations occur on documents matching the provided query/filter.
+   *
+   * @param query GROQ-filter to listen to changes for
+   * @param params Optional query parameters
+   * @param options Listener options
+   */
+  listen(query: string, params?: QueryParams): Observable<MutationEvent>
+
+  /**
+   * Set up a listener that will be notified when mutations occur on documents matching the provided query/filter.
+   *
+   * @param query GROQ-filter to listen to changes for
+   * @param params Optional query parameters
+   * @param options Listener options
+   */
+  listen(query: string, params?: QueryParams, options?: ListenOptions): Observable<ListenEvent>
+
+  /**
+   * Perform a GROQ-query against the configured dataset.
+   *
+   * @param query GROQ-query to perform
+   * @param params Optional query parameters
+   * @param options Request options
+   */
+  fetch<T extends FilteredResponseQueryOptions | UnfilteredReponseQueryOptions>(
+    query: string,
+    params?: QueryParams,
+    options?: T
+  ): Promise<T extends UnfilteredReponseQueryOptions ? RawQueryResponse : any>
+
+  /**
+   * Fetch a single document with the given ID.
+   *
+   * @param id Document ID to fetch
+   */
+  getDocument(id: string): Promise<SanityDocument | undefined>
```

Comment:
`getDocument<R = any>(id: string): Promise<SanityDocument<R> | undefined>`

Again, not sure, just making a point, that users should have means to provide the typings upfront instead of casting on response.

Reply:
This should have to cascade to all responses.

Reply:
I _think_ I got all the methods covered by the new signature, allowing you to set the return type.